dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/TrackerObjectsMapper.java

```diff
@@ -0,0 +1,348 @@
+    dbEnrollment.setProgram(program);
+    dbEnrollment.setTrackedEntity(trackedEntity);
+    dbEnrollment.setFollowup(enrollment.isFollowUp());
+    dbEnrollment.setGeometry(enrollment.getGeometry());
+
+    EnrollmentStatus previousStatus = dbEnrollment.getStatus();
+    dbEnrollment.setStatus(enrollment.getStatus());
+
+    if (previousStatus != dbEnrollment.getStatus()) {
+      switch (dbEnrollment.getStatus()) {
+        case ACTIVE -> {
+          dbEnrollment.setCompletedDate(null);
+          dbEnrollment.setCompletedBy(null);
+        }
+        case COMPLETED -> {
+          dbEnrollment.setCompletedDate(now);
+          dbEnrollment.setCompletedBy(user.getUsername());
+        }
+        case CANCELLED -> {
+          dbEnrollment.setCompletedDate(now);
+          dbEnrollment.setCompletedBy(null);
+        }
+      }
+    }
+
+    if (isNotEmpty(enrollment.getNotes())) {
+      dbEnrollment
+          .getNotes()
+          .addAll(
+              enrollment.getNotes().stream()
+                  .map(note -> map(preheat, note, user))
+                  .collect(Collectors.toSet()));
+    }
+    return dbEnrollment;
+  }
+
+  public static @Nonnull Event map(
+      @Nonnull TrackerPreheat preheat,
+      @Nonnull org.hisp.dhis.tracker.imports.domain.Event event,
+      @Nonnull UserDetails user) {
+    Event dbEvent = preheat.getEvent(event.getEvent());
+    ProgramStage programStage = preheat.getProgramStage(event.getProgramStage());
+    Program program = preheat.getProgram(event.getProgram());
+    OrganisationUnit organisationUnit = preheat.getOrganisationUnit(event.getOrgUnit());
+
+    Date now = new Date();
+
+    if (dbEvent == null) {
+      dbEvent = new Event();
+      dbEvent.setUid(event.getEvent());
+      dbEvent.setCreated(now);
+      dbEvent.setStoredBy(event.getStoredBy());
+      dbEvent.setCreatedByUserInfo(UserInfoSnapshot.from(user));
+    }
+    dbEvent.setCreatedAtClient(DateUtils.fromInstant(event.getCreatedAtClient()));
+    dbEvent.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
+    dbEvent.setLastUpdated(now);
+    dbEvent.setLastUpdatedAtClient(DateUtils.fromInstant(event.getUpdatedAtClient()));
+    dbEvent.setEnrollment(getEnrollment(preheat, event.getEnrollment(), program));
+    dbEvent.setProgramStage(programStage);
+    dbEvent.setOrganisationUnit(organisationUnit);
+    dbEvent.setOccurredDate(DateUtils.fromInstant(event.getOccurredAt()));
+    dbEvent.setScheduledDate(DateUtils.fromInstant(event.getScheduledAt()));
+
+    if (event.getAttributeOptionCombo().isNotBlank()) {
+      dbEvent.setAttributeOptionCombo(
+          preheat.getCategoryOptionCombo(event.getAttributeOptionCombo()));
+    } else {
+      dbEvent.setAttributeOptionCombo(preheat.getDefault(CategoryOptionCombo.class));
+    }
+
+    dbEvent.setGeometry(event.getGeometry());
+
+    EventStatus currentStatus = event.getStatus();
+    EventStatus previousStatus = dbEvent.getStatus();
+
```

Comment:
like above if you make a paragraph (no empty lines) in between all the status logic its easier to read this logic